File: app/core/TriageRulesLoader.php
<?php
/**
 * Clinical triage rules from triage_rules.csv
 */

final class TriageRulesLoader
{
    /** @var list<array<string, string>>|null */
    private static ?array $rules = null;

    /** @return list<array<string, string>> */
    public static function rules(): array
    {
        if (self::$rules !== null) {
            return self::$rules;
        }

        self::$rules = [];
        foreach ([
            BASE_PATH . '/data/nlp/triage_rules_cds.csv',
            BASE_PATH . '/data/nlp/triage_rules.csv',
        ] as $path) {
            self::loadRulesFromCsv($path);
        }

        usort(self::$rules, static fn (array $a, array $b): int => strlen($b['hiligaynon_pattern']) <=> strlen($a['hiligaynon_pattern']));

        return self::$rules;
    }

    private static function loadRulesFromCsv(string $path): void
    {
        if (!is_readable($path)) {
            return;
        }

        $handle = fopen($path, 'r');
        if ($handle === false) {
            return;
        }

        $header = fgetcsv($handle);
        while (($row = fgetcsv($handle)) !== false) {
            $data = array_combine(
                array_map(static fn ($h) => strtolower(trim((string) $h)), $header ?: []),
                array_map(static fn ($v) => trim((string) $v), $row)
            ) ?: [];
            $hil = strtolower((string) ($data['hiligaynon_pattern'] ?? ''));
            $eng = strtolower((string) ($data['english_pattern'] ?? ''));
            if ($hil === '' && $eng === '') {
                continue;
            }
            $status = strtolower((string) ($data['status'] ?? 'active'));
            if ($status !== '' && $status !== 'active') {
                continue;
            }
            $tri = strtolower((string) ($data['triage_level'] ?? 'routine'));
            self::$rules[] = [
                'hiligaynon_pattern' => $hil,
                'english_pattern'    => $eng,
                'triage_level'       => self::mapLevel($tri),
                'severity'           => (string) ($data['severity'] ?? 'moderate'),
                'medical_category'   => (string) ($data['medical_category'] ?? ''),
                'reason'             => (string) ($data['reason'] ?? ''),
                'source'             => basename($path),
            ];
        }
        fclose($handle);
    }

    /** @return array{triage_level:string,severity:string,reason:string,source:string,pattern:string}|null */
    public static function matchTriage(string $original, string $english = ''): ?array
    {
        $hayHil = strtolower($original);
        $hayEng = strtolower($english);
        foreach (self::rules() as $rule) {
            $hil = $rule['hiligaynon_pattern'];
            $eng = $rule['english_pattern'];
            if ($hil !== '' && str_contains($hayHil, $hil)) {
                return [
                    'triage_level' => $rule['triage_level'],
                    'severity'     => $rule['severity'],
                    'reason'       => $rule['reason'] !== '' ? $rule['reason'] : 'Matched CDS rule: ' . $hil,
                    'source'       => (string) ($rule['source'] ?? 'triage_rules.csv'),
                    'pattern'      => $hil,
                ];
            }
            if ($eng !== '' && str_contains($hayEng, $eng)) {
                return [
                    'triage_level' => $rule['triage_level'],
                    'severity'     => $rule['severity'],
                    'reason'       => $rule['reason'] !== '' ? $rule['reason'] : 'Matched CDS rule: ' . $eng,
                    'source'       => (string) ($rule['source'] ?? 'triage_rules.csv'),
                    'pattern'      => $eng,
                ];
            }
        }

        return null;
    }

    private static function mapLevel(string $tri): string
    {
        return match ($tri) {
            'non_urgent', 'routine' => 'LOW',
            'urgent'                => 'HIGH',
            'emergency', 'critical' => 'EMERGENCY',
            default                 => strtoupper($tri),
        };
    }
}

File: ai_service/triage_rules_loader.py
"""Clinical triage rules from triage_rules.csv."""

from __future__ import annotations

import csv
import re
from functools import lru_cache
from pathlib import Path
from typing import Any

_DATA = Path(__file__).resolve().parent.parent / "data" / "nlp" / "triage_rules.csv"

_LEVEL_MAP = {
    "non_urgent": "LOW",
    "routine": "LOW",
    "urgent": "HIGH",
    "emergency": "EMERGENCY",
    "critical": "EMERGENCY",
}


@lru_cache(maxsize=1)
def load_rules() -> tuple[dict[str, Any], ...]:
    if not _DATA.is_file():
        return ()
    rules: list[dict[str, Any]] = []
    with _DATA.open(encoding="utf-8", newline="") as f:
        for row in csv.DictReader(f):
            hil = (row.get("hiligaynon_pattern") or "").strip().lower()
            eng = (row.get("english_pattern") or "").strip().lower()
            if not hil and not eng:
                continue
            tri = (row.get("triage_level") or "routine").strip().lower()
            rules.append(
                {
                    "hiligaynon_pattern": hil,
                    "english_pattern": eng,
                    "triage_level": _LEVEL_MAP.get(tri, tri.upper()),
                    "severity": (row.get("severity") or "moderate").strip().lower(),
                    "medical_category": (row.get("medical_category") or "").strip(),
                    "reason": (row.get("reason") or "").strip(),
                }
            )
    rules.sort(key=lambda r: -len(r.get("hiligaynon_pattern") or ""))
    return tuple(rules)


def match_triage(original: str, english: str = "") -> dict[str, str] | None:
    hay_hil = (original or "").lower()
    hay_eng = (english or "").lower()
    for rule in load_rules():
        hil_pat = rule.get("hiligaynon_pattern") or ""
        eng_pat = rule.get("english_pattern") or ""
        if hil_pat and hil_pat in hay_hil:
            return {
                "triage_level": rule["triage_level"],
                "severity": rule["severity"],
                "reason": rule["reason"] or f"Matched triage rule: {hil_pat}",
                "source": "triage_rules.csv",
            }
        if eng_pat and eng_pat in hay_eng:
            return {
                "triage_level": rule["triage_level"],
                "severity": rule["severity"],
                "reason": rule["reason"] or f"Matched triage rule: {eng_pat}",
                "source": "triage_rules.csv",
            }
    return None


def clear_cache() -> None:
    load_rules.cache_clear()
